File: code/core/internal/api/metrics.go
package api

import (
	"net/http"
	"runtime"
	"time"
)

// SystemMetrics represents the complete system metrics response.
type SystemMetrics struct {
	Timestamp     string          `json:"timestamp"`
	Version       string          `json:"version"`
	UptimeSeconds int64           `json:"uptime_seconds"`
	Runtime       RuntimeMetrics  `json:"runtime"`
	WebSocket     WSMetrics       `json:"websocket"`
	MQTT          MQTTMetrics     `json:"mqtt"`
	KNXBridge     *KNXMetrics     `json:"knx_bridge,omitempty"`
	Devices       DeviceMetrics   `json:"devices"`
	Database      DatabaseMetrics `json:"database"`
}

// RuntimeMetrics contains Go runtime statistics.
type RuntimeMetrics struct {
	Goroutines    int     `json:"goroutines"`
	MemoryAllocMB float64 `json:"memory_alloc_mb"`
	MemoryTotalMB float64 `json:"memory_total_mb"`
	NumGC         uint32  `json:"num_gc"`
}

// WSMetrics contains WebSocket hub statistics.
type WSMetrics struct {
	ConnectedClients int `json:"connected_clients"`
}

// MQTTMetrics contains MQTT client statistics.
type MQTTMetrics struct {
	Connected bool `json:"connected"`
}

// KNXMetrics contains KNX bridge statistics.
type KNXMetrics struct {
	Connected      bool   `json:"connected"`
	Status         string `json:"status"`
	TelegramsTx    uint64 `json:"telegrams_tx"`
	TelegramsRx    uint64 `json:"telegrams_rx"`
	DevicesManaged int    `json:"devices_managed"`
}

// DeviceMetrics contains device registry statistics.
type DeviceMetrics struct {
	Total    int            `json:"total"`
	ByHealth map[string]int `json:"by_health"`
	ByDomain map[string]int `json:"by_domain"`
}

// DatabaseMetrics contains database connection pool statistics.
type DatabaseMetrics struct {
	OpenConnections int   `json:"open_connections"`
	InUse           int   `json:"in_use"`
	Idle            int   `json:"idle"`
	WaitCount       int64 `json:"wait_count"`
}

// handleMetrics returns comprehensive system metrics.
func (s *Server) handleMetrics(w http.ResponseWriter, _ *http.Request) {
	// Collect runtime stats
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	// Build metrics response
	metrics := SystemMetrics{
		Timestamp:     time.Now().UTC().Format(time.RFC3339),
		Version:       s.version,
		UptimeSeconds: int64(time.Since(s.startTime).Seconds()),
		Runtime: RuntimeMetrics{
			Goroutines:    runtime.NumGoroutine(),
			MemoryAllocMB: float64(memStats.Alloc) / 1024 / 1024,
			MemoryTotalMB: float64(memStats.TotalAlloc) / 1024 / 1024,
			NumGC:         memStats.NumGC,
		},
		WebSocket: WSMetrics{
			ConnectedClients: s.hub.ClientCount(),
		},
	}

	// MQTT metrics (if available)
	if s.mqtt != nil {
		metrics.MQTT = MQTTMetrics{
			Connected: s.mqtt.IsConnected(),
		}
	}

	// KNX bridge metrics (if available)
	if s.knxMetricsProvider != nil {
		knxStats := s.knxMetricsProvider.GetMetrics()
		metrics.KNXBridge = &KNXMetrics{
			Connected:      knxStats.Connected,
			Status:         knxStats.Status,
			TelegramsTx:    knxStats.TelegramsTx,
			TelegramsRx:    knxStats.TelegramsRx,
			DevicesManaged: knxStats.DevicesManaged,
		}
	}

	// Device registry stats
	regStats := s.registry.GetStats()
	metrics.Devices = DeviceMetrics{
		Total:    regStats.TotalDevices,
		ByHealth: make(map[string]int),
		ByDomain: make(map[string]int),
	}
	for health, count := range regStats.ByHealthStatus {
		metrics.Devices.ByHealth[string(health)] = count
	}
	for domain, count := range regStats.ByDomain {
		metrics.Devices.ByDomain[string(domain)] = count
	}

	// Database stats (if available)
	if s.db != nil {
		dbStats := s.db.Stats()
		metrics.Database = DatabaseMetrics{
			OpenConnections: dbStats.OpenConnections,
			InUse:           dbStats.InUse,
			Idle:            dbStats.Idle,
			WaitCount:       dbStats.WaitCount,
		}
	}

	writeJSON(w, http.StatusOK, metrics)
}

File: code/ui/wallpanel/lib/models/metrics.dart
/// System metrics data model for the Admin dashboard.
///
/// Maps to the /api/v1/metrics endpoint response.
class SystemMetrics {
  final String timestamp;
  final String version;
  final int uptimeSeconds;
  final RuntimeMetrics runtime;
  final WebSocketMetrics websocket;
  final MQTTMetrics mqtt;
  final KNXBridgeMetrics? knxBridge;
  final DeviceMetrics devices;
  final DatabaseMetrics database;

  SystemMetrics({
    required this.timestamp,
    required this.version,
    required this.uptimeSeconds,
    required this.runtime,
    required this.websocket,
    required this.mqtt,
    this.knxBridge,
    required this.devices,
    required this.database,
  });

  factory SystemMetrics.fromJson(Map<String, dynamic> json) {
    return SystemMetrics(
      timestamp: json['timestamp'] as String,
      version: json['version'] as String,
      uptimeSeconds: json['uptime_seconds'] as int,
      runtime: RuntimeMetrics.fromJson(json['runtime'] as Map<String, dynamic>),
      websocket:
          WebSocketMetrics.fromJson(json['websocket'] as Map<String, dynamic>),
      mqtt: MQTTMetrics.fromJson(json['mqtt'] as Map<String, dynamic>),
      knxBridge: json['knx_bridge'] != null
          ? KNXBridgeMetrics.fromJson(
              json['knx_bridge'] as Map<String, dynamic>)
          : null,
      devices: DeviceMetrics.fromJson(json['devices'] as Map<String, dynamic>),
      database:
          DatabaseMetrics.fromJson(json['database'] as Map<String, dynamic>),
    );
  }

  /// Returns uptime as a human-readable string (e.g., "2h 15m").
  String get uptimeFormatted {
    final hours = uptimeSeconds ~/ 3600;
    final minutes = (uptimeSeconds % 3600) ~/ 60;
    final seconds = uptimeSeconds % 60;

    if (hours > 0) {
      return '${hours}h ${minutes}m';
    } else if (minutes > 0) {
      return '${minutes}m ${seconds}s';
    } else {
      return '${seconds}s';
    }
  }
}

class RuntimeMetrics {
  final int goroutines;
  final double memoryAllocMB;
  final double memoryTotalMB;
  final int numGC;

  RuntimeMetrics({
    required this.goroutines,
    required this.memoryAllocMB,
    required this.memoryTotalMB,
    required this.numGC,
  });

  factory RuntimeMetrics.fromJson(Map<String, dynamic> json) {
    return RuntimeMetrics(
      goroutines: json['goroutines'] as int,
      memoryAllocMB: (json['memory_alloc_mb'] as num).toDouble(),
      memoryTotalMB: (json['memory_total_mb'] as num).toDouble(),
      numGC: json['num_gc'] as int,
    );
  }
}

class WebSocketMetrics {
  final int connectedClients;

  WebSocketMetrics({required this.connectedClients});

  factory WebSocketMetrics.fromJson(Map<String, dynamic> json) {
    return WebSocketMetrics(
      connectedClients: json['connected_clients'] as int,
    );
  }
}

class MQTTMetrics {
  final bool connected;

  MQTTMetrics({required this.connected});

  factory MQTTMetrics.fromJson(Map<String, dynamic> json) {
    return MQTTMetrics(
      connected: json['connected'] as bool,
    );
  }
}

class KNXBridgeMetrics {
  final bool connected;
  final String status;
  final int telegramsTx;
  final int telegramsRx;
  final int devicesManaged;

  KNXBridgeMetrics({
    required this.connected,
    required this.status,
    required this.telegramsTx,
    required this.telegramsRx,
    required this.devicesManaged,
  });

  factory KNXBridgeMetrics.fromJson(Map<String, dynamic> json) {
    return KNXBridgeMetrics(
      connected: json['connected'] as bool,
      status: json['status'] as String,
      telegramsTx: json['telegrams_tx'] as int,
      telegramsRx: json['telegrams_rx'] as int,
      devicesManaged: json['devices_managed'] as int,
    );
  }
}

class DeviceMetrics {
  final int total;
  final Map<String, int> byHealth;
  final Map<String, int> byDomain;

  DeviceMetrics({
    required this.total,
    required this.byHealth,
    required this.byDomain,
  });

  factory DeviceMetrics.fromJson(Map<String, dynamic> json) {
    return DeviceMetrics(
      total: json['total'] as int,
      byHealth: (json['by_health'] as Map<String, dynamic>)
          .map((k, v) => MapEntry(k, v as int)),
      byDomain: (json['by_domain'] as Map<String, dynamic>)
          .map((k, v) => MapEntry(k, v as int)),
    );
  }
}

class DatabaseMetrics {
  final int openConnections;
  final int inUse;
  final int idle;
  final int waitCount;

  DatabaseMetrics({
    required this.openConnections,
    required this.inUse,
    required this.idle,
    required this.waitCount,
  });

  factory DatabaseMetrics.fromJson(Map<String, dynamic> json) {
    return DatabaseMetrics(
      openConnections: json['open_connections'] as int,
      inUse: json['in_use'] as int,
      idle: json['idle'] as int,
      waitCount: json['wait_count'] as int,
    );
  }
}
